### PR TITLE
LibWeb: Introduce navigable and traversable

### PR DESCRIPTION
**Pull request by kalenikaliaksandr**
it's a part of https://github.com/SerenityOS/serenity/pull/18219
patches from Andreas's navigables work I need to get merge :)

### CODE REVIEW

File: Userland/Libraries/LibWeb/HTML/Navigable.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+namespace Web::HTML {
+
+// https://html.spec.whatwg.org/multipage/document-sequences.html#navigable
+class Navigable : public JS::Cell {
+    JS_CELL(Navigable, JS::Cell);
+
+public:
+    virtual ~Navigable() override;
+
+    DeprecatedString id() const;
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/Navigable.h:22]: Let's not introduce more uses of `DeprecatedString` in new code. Also should return a const ref.

File: Userland/Libraries/LibWeb/HTML/TraversableNavigable.cpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+{
+    Base::visit_edges(visitor);
+    for (auto& entry : m_session_history_entries)
+        visitor.visit(entry);
+}
+
+// https://html.spec.whatwg.org/multipage/document-sequences.html#top-level-traversable
+bool TraversableNavigable::is_top_level_traversable() const
+{
+    return parent() == nullptr;
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/TraversableNavigable.cpp:26]: `// A top-level traversable is a traversable navigable with a null parent.`

File: Userland/Libraries/LibWeb/HTML/TraversableNavigable.cpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+bool TraversableNavigable::running_nested_apply_history_step() const
+{
+    return m_running_nested_apply_history_step;
+}
+
+HTML::VisibilityState TraversableNavigable::system_visibility_state() const
+{
+    return m_system_visibility_state;
+}
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/TraversableNavigable.cpp:47]: Same question here regarding plain getters.

**Review by linusg — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibWeb/Forward.h [hunk 4]
```diff
@@ -240,7 +240,7 @@ class DOMRectReadOnly;
 
 namespace Web::HTML {
 class BrowsingContext;
-class BrowsingContextContainer;
+class NavigableContainer;
```
[the remark this answers is not in the document]
Comment (kalenikaliaksandr) [reply Userland/Libraries/LibWeb/Forward.h]: ah, it's so embarrassing messing up sorting so often, sorry.

[the remark this answers is not in the document]
Comment (linusg) [reply Userland/Libraries/LibWeb/Forward.h]: no, I'm just being unusually strict about it, most people don't care :)

File: Userland/Libraries/LibWeb/HTML/TraversableNavigable.h [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+public:
+    virtual ~TraversableNavigable() override;
+
+    bool is_top_level_traversable() const;
+
+    int current_session_history_step() const { return m_current_session_history_step; };
+    Vector<JS::NonnullGCPtr<SessionHistoryEntry>> const& session_history_entries() const { return m_session_history_entries; };
+    bool running_nested_apply_history_step() const { return m_running_nested_apply_history_step; };
+    HTML::VisibilityState system_visibility_state() const { return m_system_visibility_state; };
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/TraversableNavigable.h:27]: You can omit `HTML::` here, we're already in the namespace

File: Userland/Libraries/LibWeb/HTML/TraversableNavigable.h [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    // https://html.spec.whatwg.org/multipage/document-sequences.html#tn-session-history-entries
+    Vector<JS::NonnullGCPtr<SessionHistoryEntry>> m_session_history_entries;
+
+    // FIXME: https://html.spec.whatwg.org/multipage/document-sequences.html#tn-session-history-traversal-queue
+
+    // https://html.spec.whatwg.org/multipage/document-sequences.html#tn-running-nested-apply-history-step
+    bool m_running_nested_apply_history_step { false };
+
+    // https://html.spec.whatwg.org/multipage/document-sequences.html#system-visibility-state
+    HTML::VisibilityState m_system_visibility_state { HTML::VisibilityState::Visible };
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/TraversableNavigable.h:46]: Likewise

**Review by linusg — COMMENTED** [review]
Looks much better, two more things I missed :)

File: Userland/Libraries/LibWeb/HTML/Navigable.h [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    bool is_delaying_load_events() const { return m_delaying_load_events; };
+    void set_delaying_load_events(bool value) { m_delaying_load_events = value; };
+
+    JS::GCPtr<SessionHistoryEntry> active_session_history_entry() const { return m_active_session_history_entry; };
+    JS::GCPtr<SessionHistoryEntry> current_session_history_entry() const { return m_current_session_history_entry; };
+
+    JS::GCPtr<DOM::Document> active_document();
+    JS::GCPtr<BrowsingContext> active_browsing_context();
+    JS::GCPtr<HTML::WindowProxy> active_window_proxy();
+    JS::GCPtr<HTML::Window> active_window();
```
[the remark this answers is not in the document]
Comment (awesomekling) [reply Userland/Libraries/LibWeb/HTML/Navigable.h]: Let's add a FIXME then, saying that these should be non-null pointers eventually :)

[the remark this answers is not in the document]
Comment (kalenikaliaksandr) [reply Userland/Libraries/LibWeb/HTML/Navigable.h]: done

[the remark this answers is not in the document]
Comment (kalenikaliaksandr) [reply Userland/Libraries/LibWeb/HTML/Navigable.h]: I changed it to return non-null now. I will have to figure how to handle problem I described here https://github.com/whatwg/html/issues/9148 without null check on active document but it's ok.

[the remark this answers is not in the document]
Comment (kalenikaliaksandr) [reply Userland/Libraries/LibWeb/HTML/Navigable.h]: well, I tried to rebase my WIP navigables branch and found out that actually active_document() can return null pointer and it is valid behaviour. for example:
1. `Document::destroy()` set associated navigable's active document to nullptr on step 7 https://html.spec.whatwg.org/multipage/document-lifecycle.html#destroy-a-document
2. `node_navigable()` tries to find navigable associated with a document but and stumble on a navigable with destroyed document https://html.spec.whatwg.org/multipage/document-sequences.html#node-navigable

[the remark this answers is not in the document]
Comment (awesomekling) [reply Userland/Libraries/LibWeb/HTML/Navigable.h]: Yeah all right. We can come back to this later :)

**Pull request merged**